mcs/class/corlib/System.Runtime.CompilerServices/ConditionalWeakTable.cs

```diff
@@ -376,112 +376,21 @@ internal ICollection<TValue> Values
-			// contain entries which have been collected or removed, and it frees those up, effectively moving
-			// down all subsequent entries in the container (not in the existing container, but in a replacement).
-			// This, however, would cause the enumerator's understanding of indices to break.  So, as long as
-			// there is any outstanding enumerator, no compaction is performed.
-
-			private ConditionalWeakTable<TKey, TValue> _table; // parent table, set to null when disposed
-			private int _currentIndex = -1;                    // the current index into the container
-			private KeyValuePair<TKey, TValue> _current;       // the current entry set by MoveNext and returned from Current
-
-			public Enumerator(ConditionalWeakTable<TKey, TValue> table)
-			{
-				Debug.Assert(table != null, "Must provide a valid table");
-				Debug.Assert(Monitor.IsEntered(table._lock), "Must hold the _lock lock to construct the enumerator");
-
-				// Store a reference to the parent table and increase its active enumerator count.
-				_table = table;
-				_currentIndex = -1;
-			}
-
-			~Enumerator() { Dispose(); }
-
-			public void Dispose()
-			{
-				// Use an interlocked operation to ensure that only one thread can get access to
-				// the _table for disposal and thus only decrement the ref count once.
-				ConditionalWeakTable<TKey, TValue> table = Interlocked.Exchange(ref _table, null);
-				if (table != null)
-				{
-					// Ensure we don't keep the last current alive unnecessarily
-					_current = default;
-
-					// Finalization is purely to decrement the ref count.  We can suppress it now.
-					GC.SuppressFinalize(this);
-				}
-			}
-
-			public bool MoveNext()
-			{
-				// Start by getting the current table.  If it's already been disposed, it will be null.
-				ConditionalWeakTable<TKey, TValue> table = _table;
-				if (table != null)
-				{
-					// Once have the table, we need to lock to synchronize with other operations on
-					// the table, like adding.
-					lock (table._lock)
-					{
-						var tombstone = GC.EPHEMERON_TOMBSTONE;
-						while (_currentIndex < table.data.Length - 1)
-						{
-							_currentIndex++;
-							var currentDataItem = table.data[_currentIndex];
-							if (currentDataItem.key != null && currentDataItem.key != tombstone)
-							{
-								_current = new KeyValuePair<TKey, TValue>((TKey)currentDataItem.key, (TValue)currentDataItem.value);
-								return true;
-							}
-						}
-					}
-				}
-
-				// Nothing more to enumerate.
-				return false;
-			}
-
-			public KeyValuePair<TKey, TValue> Current
-			{
-				get
+				foreach (var eph in this.data) 
 				{
-					if (_currentIndex < 0)
-					{
-						ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumOpCantHappen();
-					}
-					return _current;
+					if (eph.key != null && eph.key != tombstone)
+						list.Add (new KeyValuePair<TKey, TValue> ((TKey) eph.key, (TValue) eph.value));
```

Comment:
This will produce a correct enumeration of key-value pair, but will break the conditional liveness offered by CWT.
IE, as long as the enumerator remains alive, nothing in the CWT will be collectible.